plugins/balancedLevelProvider.go

```diff
@@ -95,35 +105,19 @@ func validateSpread(spread float64) {
 
 // GetLevels impl.
 func (p *balancedLevelProvider) GetLevels(maxAssetBase float64, maxAssetQuote float64) ([]api.Level, error) {
-	_maxAssetBase := maxAssetBase + p.virtualBalanceBase
-	_maxAssetQuote := maxAssetQuote + p.virtualBalanceQuote
-	// represents the amount that was meant to be included in a previous level that we excluded because we skipped that level
-	amountCarryover := 0.0
-	levels := []api.Level{}
-	for i := int16(0); i < p.maxLevels; i++ {
-		level, e := p.getLevel(_maxAssetBase, _maxAssetQuote)
-		if e != nil {
-			return nil, e
-		}
-
-		// always update _maxAssetBase and _maxAssetQuote to account for the level we just calculated, ensures price moves across levels regardless of inclusion of prior levels
-		_maxAssetBase, _maxAssetQuote = updateAssetBalances(level, p.useMaxQuoteInTargetAmountCalc, _maxAssetBase, _maxAssetQuote)
+	if !p.needNewLevels {
+		log.Println("no offers were taken, leave levels as they are")
+		return p.lastLevels, nil
+	}
 
-		// always take a spread off the amountCarryover
-		amountCarryoverSpread := p.getRandomSpread(p.minAmountCarryoverSpread, p.maxAmountCarryoverSpread)
-		amountCarryover *= (1 - amountCarryoverSpread)
+	levels, e := p.recomputeLevels(maxAssetBase, maxAssetQuote)
+	if e != nil {
+		log.Fatalf("unable to generate new levels: %s\n", e)
```

Comment:
we should **never** use log.Fatal
instead we should wrap the error like this: `fmt.Errorf("unable to generate new levels: %s", e)`.
The framework will handle retries etc.